leshan-client-demo/src/main/java/org/eclipse/leshan/client/demo/LeshanClientDemo.java

```diff
@@ -807,4 +809,18 @@ public void run() {
             }
         }
     }
+
+    private static class ShutdownOnUnexpectedErrorObserver extends LwM2mClientObserverAdapter {
```

Comment:
I hope I'm not too boring with this but are you sure we should keep the backward compatibility behavior ?
As I said in https://github.com/eclipse/leshan/pull/941#discussion_r534009976, we could maybe consider the current behavior as a bug rather as a behavior which should not be changed ?

WDYT ?

Of course if you insist we can only add this behavior to `LeshanClientDemo` but it's hard to me to find why keeping a kind of process zombie alive should be a good idea.(https://github.com/eclipse/leshan/issues/933#issue-746373116)

Reply:
Yeah, this has been a point for consideration.
The previous pull-request (#941) was for the new major version 2.0, on the other hand, this pull-request aims to update the current stable version. So I'm a little bit afraid of breaking the backward compatibility...
But, as you mentioned, it is difficult to find out the reason to encourage keeping the living-dead client process.
This is just my opinion, it would be better to set the hook in `LeshanClinet`; but anyway I'd like to follow and respect the decision of this project. Let's do the same thing with #941?

Reply:
Ok so I cherrypick the commit from master instead of the one in this PR.